docs/basics/tx-lifecycle.md

```diff
@@ -87,7 +87,7 @@ When `Tx` is received by the application from the underlying consensus engine (e
 
 ### AnteHandler
 
-The [`AnteHandler`](../basics/gas-fees.md#antehandler), which is technically optional but should be defined for each application, is run. A deep copy of the internal state, `checkState`, is made and the defined `AnteHandler` performs limited checks specified for the transaction type. Using a copy allows the handler to do stateful checks for `Tx` without modifying the last committed state, and revert back to the original if the execution fails.
+The [`AnteHandler`](../basics/gas-fees.md#antehandler), which is technically optional but should be defined for each application, is run. A deep copy of the internal state, `checkState`, is made and the defined `AnteHandler` performs limited checks specified for the transaction type. Using a copy allows the AnteHandler to do stateful checks for `Tx` without modifying the last committed state, and revert back to the original if the execution fails.
```

Comment:
1. Why do we link to the `gas-fees.md` file?
2. The first sentence needs to be updated (it sounds weird).
3. _A deep copy of the internal state,_ -> I don't think we do a deep copy. We only use snapshot functionality.
4. I think `AnteHandler` is required for MsgService. cc: @aaronc

Reply:
Addressed 1 & 2. For 4, I think AnteHandler is not needed for MsgService, `runMsgs` will be run whether or not there's an antehandler.

I tried to read the code & comments for 3, it's still unclear to me whether this is true or not. Do you have an idea @robert-zaremba ?

Reply:
Copying internal storage would be very expensive (100+GB for Gaia). IIRC, the storage has few layers. You can create a aversion on top of other version, and it won't get saved until we commit. But I don't know the details - did not read the code.